### PR TITLE
OS-8178 Allow `null` for vmadm(1M) integers to mean the same as the e…

### PR DESCRIPTION
**Pull request by danmcd**
…mpty-string.

### CODE REVIEW

**Comment by danmcd** [comment]
I want to be sure I'm not breaking any assumptions.  Please consider that in your reviews.

**Comment by sjorge** [comment]
I actually toyed around with it before when looking to see why some parameters did not get cleared properly when setting to null.

I didn't run into issues then (nor did it actually fix my issues) but it's on the patch list for the next pi I build which will probably be tomorrow.

**Comment by jlevon** [comment]
shouldn't this have a test?

**Comment by danmcd** [comment]
> shouldn't this have a test?

Yes... yes it should.  Watch for it later today.

**Comment by sjorge** [comment]
I did pull it in to my new PI I build, I create manual test vm where I forced bhvye_extra_opts to be "" and once where it was null and it had the same result.

I also ran my salt state that creates a docker zone and native zone and they were fine.  The docker one is very finicky because all the weird stuff required in internal_metadata. There as o change between the old and new PI with the patch.

But yes, a test would be nice.

**Comment by danmcd** [comment]
Pardon delay, needed to run existing `/usr/vm/test` tests.  They were equal on a stock PI from late-April, and a PI with most of this change in it.  Added a test now to make it complete, but need to rebuild the PI to see if it checks out.

@sjorge --> `bhvye_extra_opts` is an *integer*?  If so, glad it worked.

@jlevon --> The new test is an old test with a slight difference.

**Comment by danmcd** [comment]
Testing with the new test in place showed no differences/regression, AND passed the new "use null" test like the "use empty string"/default test did.

**Comment by sjorge** [comment]
I found a difference! :(

```
{
  "brand": "bhyve",
  "alias": "test",
  "ram": 1024,
  "bootrom": "uefi",
  "vnc_port": -1,
  "flexible_disk_size": 2048,
  "disks": [
    {
      "boot": true,
      "model": "virtio",
      "media": "disk",
      "size": 1024,
      "compression": "off",
      "block_size": 4096
    }
  ],
  "nics": [
    {
      "vlan_id": 10,
      "nic_tag": "trunk",
      "ips": [
        "dhcp",
        "addrconf"
      ],
      "model": "virtio"
    }
  ],
  "bhyve_extra_opts": null
}
```

Will result in a vm with vcpu set to 1 (default) and empty bhyve_extra_opts (str) as expected.

This however:
```
{
  "brand": "bhyve",
  "alias": "test",
  "ram": 1024,
  "bootrom": "uefi",
  "vcpus": null,
  "vnc_port": -1,
  "flexible_disk_size": 2048,
  "disks": [
    {
      "boot": true,
      "model": "virtio",
      "media": "disk",
      "size": 1024,
      "compression": "off",
      "block_size": 4096
    }
  ],
  "nics": [
    {
      "vlan_id": 10,
      "nic_tag": "trunk",
      "ips": [
        "dhcp",
        "addrconf"
      ],
      "model": "virtio"
    }
  ],
  "bhyve_extra_opts": null
}
```

Results in vpcu being unset... instead of being the default.

pargs for bhyve also indicates that '-c 1' is missing in the first example.

```
argv[0]: bhyve
argv[1]: -H
argv[2]: -U
argv[3]: e30a0800-a582-41e0-d014-f8a940cb6e53
argv[4]: -B
argv[5]: 1,manufacturer=Joyent,product=SmartDC HVM,version=7.20200527T091055Z,serial=e30a0800-a582-41e0-d014-f8a940cb6e53,sku=001,family=Virtual Machine
argv[6]: -s
argv[7]: 31,lpc
argv[8]: -l
argv[9]: bootrom,/usr/share/bhyve/uefi-rom.bin
argv[10]: -l
argv[11]: com1,/dev/zconsole
argv[12]: -l
argv[13]: com2,socket,/tmp/vm.ttyb
argv[14]: -s
argv[15]: 0,hostbridge,model=i440fx
argv[16]: -c
argv[17]: 1
argv[18]: -m
argv[19]: 1024
argv[20]: -s
argv[21]: 0:4:0,virtio-blk,/dev/zvol/rdsk/zones/e30a0800-a582-41e0-d014-f8a940cb6e53/disk0
argv[22]: -s
argv[23]: 6:0,virtio-net-viona,net0
argv[24]: SYSbhyve-843
```

vs

```
35472:  bhyve -H -U 18566c61-9a8a-ec92-b641-d4b5ceb18f20 -B 1,manufacturer=Joyent,produ
argv[0]: bhyve
argv[1]: -H
argv[2]: -U
argv[3]: 18566c61-9a8a-ec92-b641-d4b5ceb18f20
argv[4]: -B
argv[5]: 1,manufacturer=Joyent,product=SmartDC HVM,version=7.20200527T091055Z,serial=18566c61-9a8a-ec92-b641-d4b5ceb18f20,sku=001,family=Virtual Machine
argv[6]: -s
argv[7]: 31,lpc
argv[8]: -l
argv[9]: bootrom,/usr/share/bhyve/uefi-rom.bin
argv[10]: -l
argv[11]: com1,/dev/zconsole
argv[12]: -l
argv[13]: com2,socket,/tmp/vm.ttyb
argv[14]: -s
argv[15]: 0,hostbridge,model=i440fx
argv[16]: -m
argv[17]: 1024
argv[18]: -s
argv[19]: 0:4:0,virtio-blk,/dev/zvol/rdsk/zones/18566c61-9a8a-ec92-b641-d4b5ceb18f20/disk0
argv[20]: -s
argv[21]: 6:0,virtio-net-viona,net0
argv[22]: SYSbhyve-842
```

**Comment by sjorge** [comment]
OK, setting `vcpu` to an empty string `""` has the same result. So it doesn't change behavior!
Maybe something that does deseve a followup issue to fix this eventually.

**Comment by jlevon** [comment]
Surely that difference is exactly what you'd want: if you're passing in null, you really do mean "undefined" not "the default"

**Comment by danmcd** [comment]
So is that vcpu issue a big enough blocker?  Or is this a followup issue?  Sounds like at most a followup issue to me, but I *am* biased.  :)

**Comment by jlevon** [comment]
I don't think there *is* an issue?

**Comment by danmcd** [comment]
Cool.  Is this good enough to approve, now that I have a new test in place (and AFAICT no regressions from the existing test suite)?

**Comment by jlevon** [comment]
If it's OK I'd like a little time to review it.

**Comment by sjorge** [comment]
> So is that vcpu issue a big enough blocker? Or is this a followup issue? Sounds like at most a followup issue to me, but I _am_ biased. :)

Not a block, see

> OK, setting `vcpu` to an empty string `""` has the same result. So it doesn't change behavior!
> Maybe something that does deseve a followup issue to fix this eventually.

At best a follow up issue. vmadm's manpage is not clear on the behavior.
Should setting something to null/undef/"" be remove this item complete or reset to whatever is the default.

Personally I would lean to the later, but I'm fine with either really. But might be useful to explicitly mention it in the man page. Which should be a follow up issue.

File: src/vm/tests/test-zfs-limits-null.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// this puts test stuff in global, so we need to tell jsl about that:
+/* jsl:import ../node_modules/nodeunit-plus/index.js */
+require('nodeunit-plus');
+
+VM.loglevel = 'DEBUG';
+
+var image_uuid = vmtest.CURRENT_SMARTOS_UUID;
+var trim = utils.trim;
+
+test('test create vm with snapshot limits, disabled by NULL', function (t) {
```
[the remark this answers is not in the document]
Comment (danmcd) [reply src/vm/tests/test-zfs-limits-null.js:19]: Fixed nit.

As to your larger question:  TODAY, updating using `vmadm update UUID zfs_snapshot_limit=null` will fail:

```
[root@curly (kebecloud) ~]# vmadm update ee62f3de-bca3-6478-fb4b-feec232b1e99 zfs_snapshot_limit=null
Invalid value for zfs_snapshot_limit: null:number
[root@curly (kebecloud) ~]# vmadm update ee62f3de-bca3-6478-fb4b-feec232b1e99 zfs_snapshot_limit=''
Successfully updated VM ee62f3de-bca3-6478-fb4b-feec232b1e99
[root@curly (kebecloud) ~]# vmadm update ee62f3de-bca3-6478-fb4b-feec232b1e99 zfs_snapshot_limit=undefined
Invalid value for zfs_snapshot_limit: null:number
```
Should that be fixed as well?

[the remark this answers is not in the document]
Comment (jlevon) [reply src/vm/tests/test-zfs-limits-null.js]: It's somewhat unfortunate, but also understandable: there's really no way to different 'null' from null in this context.

However, could you please add a separate paragraph under the "vmadm update" description highlighting a) how to clear out values b) how the cmdline version is different

[the remark this answers is not in the document]
Comment (danmcd) [reply src/vm/tests/test-zfs-limits-null.js]: Will do.  And funny thing, this works:
```
[root@curly (kebecloud) ~]# vmadm update ee62f3de-bca3-6478-fb4b-feec232b1e99 '{"zfs_snapshot_limit": null}'
Successfully updated VM ee62f3de-bca3-6478-fb4b-feec232b1e99
You have new mail in /var/mail/root
[root@curly (kebecloud) ~]# 
```

[the remark this answers is not in the document]
Comment (jlevon) [reply src/vm/tests/test-zfs-limits-null.js]: I'm confused; your example shows it working

File: src/vm/man/vmadm.1m.md [hunk 2]
```diff
@@ -390,6 +390,15 @@ tab-complete UUIDs rather than having to type them out for every command.
         JSON object on stdin (though it will refuse to work if stdin is a tty),
         or pass property=value arguments on the command line.
 
+	NOTE that property=value arguments may behave differently than JSON
```
Comment (jlevon) [inline src/vm/man/vmadm.1m.md:393]: This uses tabs whereas the rest is space indented.

Pending the other thing about passing json directly on the cmdline, I'd reword this as something like

```
Many properties can be cleared by specifying their value as null in the JSON, e.g.

  { ... "zfs_snapshot_limit": null }

However this doesn't work via a direct `vmadm update prop=null` command
